src/UI/Actions.hs

```diff
@@ -862,316 +884,333 @@ abort = Action ["cancel"] (reset (Proxy :: Proxy v) (Proxy :: Proxy a))
+setTags :: [TagOp] -> Action v ctx ()
 setTags ops =
     Action
     { _aDescription = ["apply given tags"]
-    , _aAction = \s -> case focusedViewWidget s of
-          ListOfMails -> selectedItemHelper (asMailIndex . miListOfMails) s (manageMailTags s ops)
-          _ -> selectedItemHelper (asMailIndex . miListOfThreads) s (manageThreadTags s ops)
+    , _aAction = do
+        w <- focusedViewWidget <$> get
+        case w of
+          ListOfMails -> selectedItemHelper (asMailIndex . miListOfMails) (manageMailTags ops)
+          _ -> selectedItemHelper (asMailIndex . miListOfThreads) (manageThreadTags ops)
     }
 
 -- | Reloads the list of threads by executing the notmuch query given
 -- by the search widget.
 --
-reloadList :: Action 'Threads 'ListOfThreads AppState
+reloadList :: Action 'Threads 'ListOfThreads ()
 reloadList = Action ["reload list of threads"] applySearch
 
 -- | Selects the next unread mail in a thread.
 --
-selectNextUnread :: Action 'ViewMail 'ListOfMails AppState
-selectNextUnread =
-  Action { _aDescription = ["select next unread"]
-         , _aAction = \s ->
-           let
-             -- find by unread tag...
-             p = Notmuch.hasTag (view (asConfig . confNotmuch . nmNewTag) s)
-             -- but if there is no resulting selection, move to the
-             -- last element in the list
-             f l = maybe (L.listMoveTo (-1) l) (const l) (view L.listSelectedL l)
-           in pure $ over (asMailIndex . miListOfMails) (f . L.listFindBy p) s
-         }
+selectNextUnread :: Action 'ViewMail 'ListOfMails ()
+selectNextUnread = Action
+  { _aDescription = ["select next unread"]
+  , _aAction = do
+      -- find by unread tag...
+      p <- uses (asConfig . confNotmuch . nmNewTag) Notmuch.hasTag
+      -- but if there is no resulting selection, move to the
+      -- last element in the list
+      let f l = maybe (L.listMoveTo (-1) l) (const l) (view L.listSelectedL l)
+      modifying (asMailIndex . miListOfMails) (f . L.listFindBy p)
+  }
 
 -- | Selects a list item. Currently only used in the file browser to
 -- select a file for attaching.
 --
-toggleListItem :: Action v 'ListOfFiles AppState
+toggleListItem :: Action v 'ListOfFiles ()
 toggleListItem =
     Action
     { _aDescription = ["toggle selected state of a list item"]
-    , _aAction = \s ->
-                      maybe
-                          (pure s)
-                          (\i -> pure $ over (asFileBrowser . fbEntries . L.listElementsL . ix i . _1) not s)
-                          (view (asFileBrowser . fbEntries . L.listSelectedL) s)
+    , _aAction = use (asFileBrowser . fbEntries . L.listSelectedL) >>= traverse_ f
     }
+    where
+      f i = modifying (asFileBrowser . fbEntries . L.listElementsL . ix i . _1) not
 
 -- | Delete an attachment from a mail currently being composed.
 --
-delete :: Action 'ComposeView 'ComposeListOfAttachments AppState
+-- TODO: could this be generalised over a type class, e.g.:
```

Comment:
This one as well into an issue with an easy fix?

Reply:
Yes I'll file an issue for it.